Python/gc_free_threading.c

```diff
@@ -1,9 +1,1660 @@
+#define GENERATION_AUTO (-1)
+
+// A linked-list of objects using the `ob_tid` field as the next pointer.
+struct worklist {
+    uintptr_t head;
+};
+
+struct worklist_iter {
+    uintptr_t *ptr;   // pointer to current object
+    uintptr_t *next;  // next value of ptr
+};
+
+struct visitor_args {
+    size_t offset;  // offset of PyObject from start of block
+};
+
+// Per-collection state
+struct collection_state {
+    struct visitor_args base;
+    PyInterpreterState *interp;
+    GCState *gcstate;
+    Py_ssize_t collected;
+    Py_ssize_t uncollectable;
+    struct worklist unreachable;
+    struct worklist legacy_finalizers;
+    struct worklist wrcb_to_call;
+    struct worklist objs_to_decref;
+};
+
+// iterate over a worklist
+#define WORKSTACK_FOR_EACH(stack, op) \
+    for ((op) = (PyObject *)(stack)->head; (op) != NULL; (op) = (PyObject *)(op)->ob_tid)
+
+// iterate over a worklist with support for removing the current object
+#define WORKSTACK_FOR_EACH_ITER(stack, iter, op) \
+    for (worklist_iter_init((iter), &(stack)->head), (op) = (PyObject *)(*(iter)->ptr); \
+         (op) != NULL; \
+         worklist_iter_init((iter), (iter)->next), (op) = (PyObject *)(*(iter)->ptr))
+
+static void
+worklist_push(struct worklist *worklist, PyObject *op)
+{
+    assert(op->ob_tid == 0);
+    op->ob_tid = worklist->head;
+    worklist->head = (uintptr_t)op;
+}
+
+static PyObject *
+worklist_pop(struct worklist *worklist)
+{
+    PyObject *op = (PyObject *)worklist->head;
+    if (op != NULL) {
+        worklist->head = op->ob_tid;
+        op->ob_tid = 0;
+    }
+    return op;
+}
+
+static void
+worklist_iter_init(struct worklist_iter *iter, uintptr_t *next)
+{
+    iter->ptr = next;
+    PyObject *op = (PyObject *)*(iter->ptr);
+    if (op) {
+        iter->next = &op->ob_tid;
+    }
+}
+
+static void
+worklist_remove(struct worklist_iter *iter)
+{
+    PyObject *op = (PyObject *)*(iter->ptr);
+    *(iter->ptr) = op->ob_tid;
+    op->ob_tid = 0;
+    iter->next = iter->ptr;
+}
+
+static inline int
+gc_is_unreachable(PyObject *op)
+{
+    return (op->ob_gc_bits & _PyGC_BITS_UNREACHABLE) != 0;
+}
+
+static void
+gc_set_unreachable(PyObject *op)
+{
+    op->ob_gc_bits |= _PyGC_BITS_UNREACHABLE;
+}
+
+static void
+gc_clear_unreachable(PyObject *op)
+{
+    op->ob_gc_bits &= ~_PyGC_BITS_UNREACHABLE;
+}
+
+static inline Py_ssize_t
+gc_get_refs(PyObject *op)
+{
+    return (Py_ssize_t)op->ob_tid;
+}
+
+static inline void
+gc_add_refs(PyObject *op, Py_ssize_t refs)
+{
+    assert(_PyObject_GC_IS_TRACKED(op));
+    op->ob_tid += refs;
+}
+
+static inline void
+gc_decref(PyObject *op)
+{
+    op->ob_tid -= 1;
+}
+
+static void
+merge_refcount(PyObject *op, Py_ssize_t extra)
+{
+    Py_ssize_t refcount = Py_REFCNT(op);
+    refcount += extra;
+
+#ifdef Py_REF_DEBUG
+    _Py_AddRefTotal(_PyInterpreterState_GET(), extra);
+#endif
+
+    op->ob_tid = 0;
+    op->ob_ref_local = 0;
+    op->ob_ref_shared = _Py_REF_SHARED(refcount, _Py_REF_MERGED);
```

Comment:
The other threads in the interpreter must be paused, so no atomics necessary. I'll add an assert here.

We could probably use `_Py_ExplicitMergeRefcount()`, but it seems convenient to avoid the atomic operations.